src/pages/o/[orgId]/surveys/[surveyId]/index.tsx

```diff
@@ -1,32 +1,346 @@
+  const { req, res } = ctx;
   const { surveyId, orgId } = ctx.params!;
+  let status: FormStatus = 'editing';
+
+  const apiClient = new BackendApiClient(req.headers);
+  let survey: ZetkinSurveyExtended;
+  try {
+    survey = await apiClient.get<ZetkinSurveyExtended>(
+      `/api/orgs/${orgId}/surveys/${surveyId}`
+    );
+  } catch (e) {
+    return { notFound: true };
+  }
+
+  if (req.method === 'POST') {
+    const form = await parseRequest(req);
+    const responses: Record<
+      string,
+      {
+        options?: number[];
+        question_id: number;
+        response?: string;
+      }
+    > = {};
+
+    for (const name in form) {
+      const isSignature = name.startsWith('sig');
+      const isPrivacy = name.startsWith('privacy');
+      const isMetadata = isSignature || isPrivacy;
+      if (isMetadata) {
+        continue;
+      }
+
+      const fields = name.split('.');
+      const questionId = fields[0];
+      const questionType = fields[1];
+
+      if (typeof responses[questionId] === 'undefined') {
+        responses[questionId] = {
+          question_id: parseInt(fields[0]),
+        };
+      }
+
+      if (questionType == 'options') {
+        if (Array.isArray(form[name])) {
+          responses[questionId].options = (form[name] as string[]).map((o) =>
+            parseInt(o)
+          );
+        } else {
+          responses[questionId].options = [parseInt((form[name] as string)!)];
+        }
+      } else if (questionType == 'text') {
+        responses[questionId].response = form[name] as string;
+      }
+    }
+
+    let signature: null | {
+      email: string;
+      first_name: string;
+      last_name: string;
+    } = null;
+    // TODO: handle other signature types
+    if (form.sig == 'email') {
+      signature = {
+        email: form['sig.email'] as string,
+        first_name: form['sig.first_name'] as string,
+        last_name: form['sig.last_name'] as string,
+      };
+    }
+
+    try {
+      await apiClient.post(
+        `/api/orgs/${orgId}/surveys/${surveyId}/submissions`,
+        {
+          responses: Object.values(responses),
+          signature,
+        }
+      );
+    } catch (e) {
+      status = 'error';
+    }
+    if (status !== 'error') {
+      status = 'submitted';
+    }
+
+    if (status === 'submitted') {
+      res.writeHead(302, {
+        Location: `/o/${orgId}/surveys/${surveyId}/submitted`,
+      });
+      res.end();
+    }
+  }
 
   return {
     props: {
       orgId,
-      surveyId,
+      status,
+      survey,
     },
   };
 }, scaffoldOptions);
 
 type PageProps = {
   orgId: string;
-  surveyId: string;
+  status: FormStatus;
+  survey: ZetkinSurveyExtended;
 };
 
-const Page: FC<PageProps> = ({ orgId, surveyId }) => {
+type FormStatus = 'editing' | 'invalid' | 'error' | 'submitted';
+type SignatureOption = 'authenticated' | 'email' | 'anonymous';
+
+function RadioFormControlLabel(props: FormControlLabelProps) {
```

Comment:
I'm not sure I understand the purpose of this component but I am sure there is one, so I would love to hear your thoughts (or a comment in the code). Is this a component that deserves to exist standalone?

Reply:
Yeah taking me a second to unpick the purpose of this one too. Been trying deleting bits and pieces of it to see what the impact is on rendering outcomes. Going to leave this one until a bit later I think.

Reply:
Deleted as of 933c9dcc6d589f6905911cbcb9499a7a1de983cf. I think some of the flow of control became a bit contorted here in the fun of the code camp but it's a lot more linear and simple now.